python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -1,16 +1,94 @@
-from typing import Any, Callable, Optional, Dict
+import inspect
+import os
+import shutil
+import tempfile
+from distutils.version import LooseVersion
+from pathlib import Path
+from typing import Any, Callable, Dict, Optional, Type
 
-from transformers.trainer import Trainer
-from torch.utils.data import Dataset as TorchDataset
+import torch
+import transformers
+import transformers.trainer
+import ray.cloudpickle as cpickle
+from torch.utils.data import DataLoader, Dataset as TorchDataset
+from transformers.training_args import TrainingArguments
 
-from ray.train.torch import TorchConfig
-from ray.ml.trainer import GenDataset
-from ray.ml.train.integrations.torch import TorchTrainer
-from ray.ml.config import ScalingConfig, RunConfig
-from ray.ml.preprocessor import Preprocessor
+from ray import train
+from ray import tune
+from ray.util import PublicAPI, get_node_ip_address
 from ray.ml.checkpoint import Checkpoint
-from ray.util import PublicAPI
-from ray.ml.constants import TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY
+from ray.ml.config import RunConfig, ScalingConfig
+from ray.ml.constants import EVALUATION_DATASET_KEY, PREPROCESSOR_KEY, TRAIN_DATASET_KEY
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.train.integrations.torch import TorchTrainer
+from ray.ml.trainer import GenDataset
+from ray.ml.utils.huggingface_utils import (
+    CHECKPOINT_PATH_ON_NODE_KEY,
+    NODE_IP_KEY,
+    process_dataset_for_hf,
+    TrainReportCallback,
+)
+from ray.train.checkpoint import TuneCheckpointManager
+from ray.train.constants import TUNE_CHECKPOINT_ID
+from ray.train.session import get_session
+from ray.train.torch import TorchConfig
+from ray.tune.utils.file_transfer import delete_on_node, sync_dir_between_nodes
+
+# This trainer uses a special checkpoint syncing logic.
+# Because HF checkpoints are very large dirs (at least several GBs),
+# we use directory checkpoints that are synced between nodes when
+# required instead of serializing the checkpoints and sending
+# bytes over nodes. This is a much more performant solution for
+# large directory checkpoints. The current implementation
+# is special for HuggingFaceTrainer, but can and should be
+# made generic.
+# TODO(ml-team): Make dir syncing checkpoint logic generic.
+
+
+# TODO(team-ml): Refactor checkpoint management along with Tune.
+class _DataParallelSyncingCheckpointManager(TuneCheckpointManager):
+    """Same as _DataParallelCheckpointManager, but syncs the dir instead
+    of serializing it."""
+
+    def add_tune_checkpoint_id(self, path: str):
+        # Store the checkpoint_id in the file so that the Tune trial can be
+        # resumed after failure or cancellation.
+        with open(Path(path).joinpath(TUNE_CHECKPOINT_ID), "w") as f:
```

Comment:
We want to avoid converting the checkpoint to dict in the first place, as that will require the entire dictionary to be loaded, converted to bytes, then have the ID added, and converted back to directory again.

Reply:
The checkpoint starts off as a dict right? So we can just call `self.add_tune_checkpoint_id` while it's still a dict, before it ever gets converted to a directory, and then not have to override the CheckpointManager API in the subclass.

Reply:
The dict itself contains the ip and path. It is used only for syncing and is not persisted.

Reply:
Correct. What I'm saying is that the actual persisting of the Tune checkpoint id can be done inside `write_checkpoint`. We don't need to override `add_tune_checkpoint_id` to do so. This way the interfaces for CheckpointManager are not being changed (`add_tune_checkpoint_id` can continue to follow the same implementation as the superclass, and work entirely with Dict checkpoints).

Reply:
Ok, gotcha